src/sql/workbench/contrib/notebook/browser/notebook.component.ts

```diff
@@ -89,6 +91,71 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 	public previewFeaturesEnabled: boolean = false;
 	public doubleClickEditEnabled: boolean;
 
+	@HostListener('document:keydown', ['$event'])
+	handleKeyboardEvent(event) {
+		const toolbarComponent = (<CellToolbarComponent>this.cellToolbar.first);
+		const toolbarEditCellAction = toolbarComponent.getEditCellAction();
+		let cellEditorProvider = this.cellEditors.find(e => e.cellGuid() === this._model.activeCell.cellGuid);
+		let is_selected = false;
+		let editorControl: CodeEditorWidget;
+		if (cellEditorProvider) {
+			let editor = cellEditorProvider.getEditor() as QueryTextEditor;
+			if (editor) {
+				editorControl = editor.getControl() as CodeEditorWidget;
+				is_selected = editorControl.hasTextFocus();
+			}
+		}
+		switch (event.key) {
+			case 'Escape':
+				if (this._model.activeCell.isEditMode) {
+					this._model.activeCell.isEditMode = !this._model.activeCell.isEditMode;
+				}
+				if (toolbarEditCellAction.editMode) {
+					toolbarEditCellAction.editMode = !toolbarEditCellAction.editMode;
+				}
+				if (editorControl) {
+					editorControl.unfocus();
```

Comment:
Do we really want to unfocus here versus setting the focus to something else that makes sense? Do we even need to unfocus if we're exiting edit mode? It seems like it'd be kind of annoying to exit edit mode and then have to tab back into the Notebook to continue on where you were.

Reply:
So unfocus over here unfocuses the editor and not the cell. The reason we need this is because even if the code cell is in non edit mode, the editor is for some reason still in edit mode and needs to be unfocused to change it to non-edit mode.
The cell is still selected when this is done, so no need to press tab.